session.go

```diff
@@ -31,182 +32,131 @@ const (
-// starting are returned on Attach().  The session will close itself if nothing
-// is attached for the duration of the session timeout.
-func NewSession(id string, command *Command, execer Execer, options *Options) *Session {
+// NewSession sets up a new session.  Any errors with starting are returned on
+// Attach().  The session will close itself if nothing is attached for the
+// duration of the session timeout.
+func NewSession(command *Command, execer Execer, options *Options) *Session {
 	tempdir := filepath.Join(os.TempDir(), "coder-screen")
 	s := &Session{
-		command:    command,
-		cond:       sync.NewCond(&sync.Mutex{}),
-		configFile: filepath.Join(tempdir, "config"),
-		execer:     execer,
-		options:    options,
-		state:      StateStarting,
-		socketsDir: filepath.Join(tempdir, "sockets"),
+		attachTimeout: 30 * time.Second,
+		command:       command,
+		cond:          sync.NewCond(&sync.Mutex{}),
+		configFile:    filepath.Join(tempdir, "config"),
+		execer:        execer,
+		id:            uuid.NewString(),
+		options:       options,
+		state:         StateStarting,
+		socketsDir:    filepath.Join(tempdir, "sockets"),
 	}
-	go s.lifecycle(id)
+	go s.lifecycle()
 	return s
 }
 
 // lifecycle manages the lifecycle of the session.
-func (s *Session) lifecycle(id string) {
-	// When this context is done the process is dead.
-	ctx, cancel := context.WithCancel(context.Background())
-
-	// Close the session down immediately if there was an error.
-	process, err := s.start(ctx, id)
+func (s *Session) lifecycle() {
+	err := s.ensureSettings()
 	if err != nil {
-		defer cancel()
-		s.setState(StateDone, xerrors.Errorf("process start: %w", err))
+		s.setState(StateDone, xerrors.Errorf("ensure settings: %w", err))
 		return
 	}
 
-	// Close the session after a timeout.  Timeouts created with AfterFunc can be
-	// reset; we will reset it as long as there are active connections.  The
-	// initial timeout for starting up is set here and will probably be less than
-	// the session timeout in most cases.  It should be at least long enough for
-	// screen to be able to start up.
-	s.timer = time.AfterFunc(30*time.Second, s.Close)
+	// The initial timeout for starting up is set here and will probably be far
+	// shorter than the session timeout in most cases.  It should be at least long
+	// enough for the first screen attach to be able to start up the daemon.
+	s.timer = time.AfterFunc(s.attachTimeout, s.Close)
 
-	// Emit the done event when the process exits.
-	go func() {
-		defer cancel()
-		err := process.Wait()
-		if err != nil {
-			err = xerrors.Errorf("process exit: %w", err)
-		}
-		s.setState(StateDone, err)
-	}()
-
-	// Handle the close event by killing the process.
+	// Handle the close event by asking screen to quit the session.  We have no
+	// way of knowing when the daemon process dies so the Go side will not get
+	// cleaned up until the timeout if the process gets killed externally (for
+	// example via `exit`).
```

Comment:
I don't think this needs to be on a separate goroutine any more. Before it made some sense that we kicked off goroutines to wait for the process to exit and also wait for close, but now we only do the latter, so it can just happen on the same goroutine as the rest of `lifecycle()` after the `setState(StateReady, nil)` call.

Reply:
Good point!!